index.bs

```diff
@@ -693,34 +849,69 @@ urlPrefix: https://w3c.github.io/FileAPI/#dfn-; type: dfn;
 		<h4 method for="Clipboard">write(|data|)</h4>
 		The {{Clipboard/write(data)}} method must run these steps:
 
-			1. Let |p| be a new [=Promise=].
+			1. Let |realm| be [=this=]'s [=relevant realm=].
+
+			1. Let |p| be [=a new promise=] in |realm|.
 
 			1. Run the following steps [=in parallel=]:
 
 				1. Let |r| be the result of running [=check clipboard write permission=] [=in parallel=]
 
-				1. If |r| is not "granted", then reject |p| with a "NotAllowedError" DOMException
+				Note: Clipboard permission is not supported on Safari. However, the read() method must be called inside
+				a user gesture event and the user must select the paste option from the native context menu that pops up
+				when read() is called from JS, otherwise, the promise will be rejected.
+
+				1. [=queue a global task=] on the [=permission task source=], given |realm|'s [=Realm/global object=], to perform the below steps:
+				
+					1. If |r| is not "granted", then reject |p| with {{"NotAllowedError"}} DOMException in |realm|.
+
+				1. [=queue a global task=] on the [=clipboard task source=], given |realm|'s [=Realm/global object=], to perform the below steps:
+
+					1. Let |itemList| and |cleanItemList| be an empty [=sequence=]&lt;{{Blob}}&gt;.
+
+					1. Let |clipboardItemList| be an empty [=clipboard items=].
+
+					1. For each |clipboardItem| in |data|:
+
+						1. For each |item| in |clipboardItem|:
```

Comment:
Here `|clipboardItem|` is a `{{ClipboardItem}}`, not a list. So you cannot loop over it.

Maybe you want to loop over `|clipboardItem|'s [=ClipboardItem/clipboard item=]'s [=list of representations=]`? But that doesn't make much sense because there's no promises in the list of representations... I'm not sure what you're trying to do here.

Reply:
`|clipboardItem|` here is `|data|` which is a `{{ClipboardItems}}` defined in `Clipboard Interface` webidl.

So, during `write()` I want to access the `Blob` or `DOMString` data (depending on what the authors provided in `ClipboardItemData`) after all the promises to `Blobs` or `DOMString` are resolved. `ClipboardItemData` is the type that has promise to `ClipboardItemDataType`. I guess this will be part of the `ClipboardItem` constructor so I can maybe assume that the promises are resolved and the `Blob` data are present? That way I can just access the `|clipboardItem|'s [=ClipboardItem/clipboard item=]'s [=list of representations=]`?

Reply:
Oh, I see, I think I misread the spec. A representation's data will be a promise, after all; the constructor step 4.5 sets it to |value|, which is a promise.

So yes, you should access |clipboardItem|'s [=ClipboardItem/clipboard item=]'s [=list of representations=]. But then for each representation, you need to "unwrap" its data. To do that, probably the best thing to do is use https://webidl.spec.whatwg.org/#dfn-perform-steps-once-promise-is-settled

Reply:
Ack.